File: linkedList/147-Insertion Sort List.js
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var insertionSortList = function (head) {
    if (!head || !head.next) return head;

    let dummy = new ListNode(0);
    dummy.next = head;
    while (head && head.next) {
        if (head.next.val >= head.val) {
            head = head.next;
            continue;
        }
        let pre = dummy;

        while (head.next.val > pre.next.val) pre = pre.next;

        let cur = head.next;
        head.next = cur.next;
        cur.next = pre.next;
        pre.next = cur;
    }
    return dummy.next;
};

File: Python/147-insertionSortList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        dummy = ListNode(float('-inf'))
        cur = head
        while cur:
            pre = dummy
            while pre.next and pre.next.val <= cur.val:
                pre = pre.next
            tmp = cur.next
            cur.next = pre.next
            pre.next = cur
            cur = tmp
        return dummy.next
